pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -427,38 +534,179 @@ func (l *ListOptionIndexer) validateColumn(column string) error {
 }
 
 // buildORClause creates an SQLite compatible query that ORs conditions built from passed filters
-func (l *ListOptionIndexer) buildORClauseFromFilters(orFilters OrFilter) (string, []any, error) {
-	var orWhereClause string
+func (l *ListOptionIndexer) buildORClauseFromFilters(orFilters OrFilter, dbName string) (string, []any, error) {
 	var params []any
+	clauses := make([]string, 0, len(orFilters.Filters))
+	var newParams []any
+	var newClause string
+	var err error
 
-	for index, filter := range orFilters.Filters {
-		opString := "LIKE"
-		if filter.Op == NotEq {
-			opString = "NOT LIKE"
+	for _, filter := range orFilters.Filters {
+		if isLabelFilter(&filter) {
+			newClause, newParams, err = l.getLabelFilter(filter, dbName)
+		} else {
+			newClause, newParams, err = l.getFieldFilter(filter)
 		}
-		columnName := toColumnName(filter.Field)
-		if err := l.validateColumn(columnName); err != nil {
+		if err != nil {
 			return "", nil, err
 		}
+		clauses = append(clauses, newClause)
+		params = append(params, newParams...)
+	}
+	return strings.Join(clauses, " OR "), params, nil
+}
 
-		orWhereClause += fmt.Sprintf(`f."%s" %s ? ESCAPE '\'`, columnName, opString)
-		format := strictMatchFmt
+//type Filter struct {
+//	Field   []string
+//	Match   []string
+//	Op      Op: one of
+//	Partial bool
+//}
+
+// Possible ops from the k8s parser:
+// KEY = and == (same) VALUE
+// KEY != VALUE
+// KEY exists []  # ,KEY, => this filter
+// KEY ! []  # ,!KEY, => assert KEY doesn't exist
+// KEY in VALUES
+// KEY notin VALUES
+
+func (l *ListOptionIndexer) getFieldFilter(filter Filter) (string, []any, error) {
+	opString := ""
+	escapeString := ""
+	columnName := toColumnName(filter.Field)
+	if err := l.validateColumn(columnName); err != nil {
+		return "", nil, err
+	}
+	switch filter.Op {
+	case Eq:
 		if filter.Partial {
-			format = matchFmt
-		}
-		match := filter.Match
-		// To allow matches on the backslash itself, the character needs to be replaced first.
-		// Otherwise, it will undo the following replacements.
-		match = strings.ReplaceAll(match, `\`, `\\`)
-		match = strings.ReplaceAll(match, `_`, `\_`)
-		match = strings.ReplaceAll(match, `%`, `\%`)
-		params = append(params, fmt.Sprintf(format, match))
-		if index == len(orFilters.Filters)-1 {
-			continue
+			opString = "LIKE"
+			escapeString = escapeBackslashDirective
+		} else {
+			opString = "="
+		}
+		clause := fmt.Sprintf(`f."%s" %s ?%s`, columnName, opString, escapeString)
+		return clause, []any{formatMatchTarget(filter)}, nil
+	case NotEq:
+		if filter.Partial {
+			opString = "NOT LIKE"
+			escapeString = escapeBackslashDirective
+		} else {
+			opString = "!="
+		}
+		clause := fmt.Sprintf(`f."%s" %s ?%s`, columnName, opString, escapeString)
+		return clause, []any{formatMatchTarget(filter)}, nil
+	case Exists:
+		clause := fmt.Sprintf(`f."%s" IS NOT NULL`, columnName)
+		return clause, []any{}, nil
+	case NotExists:
+		clause := fmt.Sprintf(`f."%s" IS NULL`, columnName)
```

Comment:
I fear this will never happen, because addIndexedField will use empty string instead of proper `NULL`:

https://github.com/rancher/lasso/pull/117/files?w=1#diff-910681530ceac7cfe911fb307754542734adee41c8d009d795aabe0d9e3803b2R200

To do this right we have to:
 - really insert `NULL`
 - check for nullability every time we do another comparision (eg. `=` or `<>`, because [3-valued logic](https://en.wikipedia.org/wiki/Null_(SQL)#Comparisons_with_NULL_and_the_three-valued_logic_(3VL)))
 - make sure this does not break the UI

Before implementing that, though, do we need new operator support in the UI at all?

IIUC:
 - Before all we had for fields was equality and inequality
 - This PR's main intent is about adding support for filtering by labels and with more operators (in, not in, exists, doesn't exist)
 - This code section basically adds the same operators for fields as well

@ericpromislow is that correct?
@richard-cox do we need them?

Reply:
Now that I think about it, my guess is that NULL/!NULL don't make sense for the non-labels, because either we index them (which means they can't be NULL, because they always come with a value, right?) or we don't index them, which means you can't query on those fields.

But NULL/!NULL do make sense for labels, because you're basically querying on the existence of a key in a hash. Also it took me a couple of hours to work out the syntax for `NotExists(label)` so I don't want to toss it that quickly.

Reply:
Thinking about yaml an empty label value `foo: ` is provided as `foo: ""` via api (json). Not sure that makes any difference?

I couldn't find anything specific in the label selector for 'without this label' / 'with a label but NULL', though the generic utility of filter by everything that doesn't have a property value might be useful at some point

Reply:
Number 1, let's focus back on fields (not labels) in this thread - discussion on labels, if needed, will be in a separate thread.

Number 2, @ericpromislow:

> my guess is that NULL/!NULL don't make sense for the non-labels, because either we index them (which means they can't be NULL, because they always come with a value, right?) or we don't index them, which means you can't query on those fields.

Hmm, my understanding is different. Assuming that we index a certain resource field, its value can very well be null (ie. missing or as explicit `null` in the YAML) - lots of structs corresponding to Kubernetes resources come with the "omitempty" directive for that reason.

We have to keep in mind that, right now, in such cases we simply [throw away the null and exchange it with an empty string](https://github.com/ericpromislow/lasso/blob/2f0f76ae5d300c21b1ae959a084256bb953984a0/pkg/cache/sql/informer/listoption_indexer.go#L197). That is arguably not really awesome, but it suffices when all we care for is string search, which is probably why this shortcut was originally taken. Now...


Number 3, @richard-cox: do I understand correctly that, from a UI perspective, we do not really need anything other than exact/inexact text matching for non-label properties, at least for now?

Reply:
@ericpromislow as clarified in https://github.com/rancher/lasso/pull/117#discussion_r1869610527, we do not really need NULL/NOT NULL filtering for non-label fields for now.

How about just not supporting that operator and returning an error, until we decide it's actually needed and then we add support for it?

Reply:
Sure, this is now pushed for both lasso and steve (steve parses, lasso complains if a filter object somehow reaches it calling for a non-label existence check, but this shouldn't happen in the field).